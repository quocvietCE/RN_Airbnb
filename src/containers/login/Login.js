import React, {Component} from 'react';
import {PropTypes} from 'prop-types';
import InputField from '../../components/form/InputField';
import Icon from 'react-native-vector-icons/FontAwesome';
import {View, Text, ScrollView, KeyboardAvoidingView} from 'react-native';
import NextArrowButton from '../../components/buttons/NextArrowButton';
import NavBarButton from '../../components/buttons/NavBarButton';
import Notification from '../../components/Notification';
import {colors, stylesApp} from '../../themes';
import styles from './styles';
// import Loader from '../../components/Loader';
import {propTypesNavigation} from '../../config/constants/defaultPropTypes';
Icon.loadFont();

class Login extends Component {
  static navigationOptions = ({navigation}) => ({
    headerRight: () => (
      <NavBarButton
        location="right"
        color={colors.white}
        text="Forgot Password"
        handleButtonPress={() => navigation.navigate('ForgotPasswordStack')}
      />
    ),
    headerLeft: () => (
      <NavBarButton
        location="left"
        icon={<Icon name="angle-left" color={colors.white} size={30} />}
        handleButtonPress={() => navigation.pop()}
      />
    ),
    headerStyle: stylesApp.transparentHeaderStyle,
    headerTransparent: true,
    headerTintColor: colors.white,
    headerTitle: '',
  });

  constructor(props) {
    super(props);
    this.state = {
      formValid: true,
      validEmail: false,
      emailAddress: '',
      password: '',
      validPassword: false,
      loadingVisible: false,
      secureInput:
        props.inputType === 'text' || props.inputType === 'email'
          ? false
          : true,
    };
    this.handlePasswordChange = this.handlePasswordChange.bind(this);
    this.toggleNextButtonState = this.toggleNextButtonState.bind(this);
    this.handleEmailChange = this.handleEmailChange.bind(this);
    this.handleNextButton = this.handleNextButton.bind(this);
    this.handleCloseNotification = this.handleCloseNotification.bind(this);
  }

  handleNextButton() {
    this.setState({loadingVisible: true});
    const {navigate} = this.props.navigation;
    const {emailAddress, password} = this.state;
    this.props.login(emailAddress, password);
    if (this.props.loggedInState) {
      this.setState({formValid: true, loadingVisible: false});
      navigate('TurnOnNotificationStack');
    } else {
      this.setState({formValid: false, loadingVisible: false});
    }
  }

  handleEmailChange(email) {
    // eslint-disable-next-line
    const emailCheckRegex = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/;
    const {validEmail} = this.state;
    this.setState({emailAddress: email});

    if (!validEmail) {
      if (emailCheckRegex.test(email)) {
        this.setState({validEmail: true});
      }
    } else if (!emailCheckRegex.test(email)) {
      this.setState({validEmail: false});
    }
  }

  handlePasswordChange(password) {
    const {validPassword} = this.state;

    this.setState({password});

    if (!validPassword) {
      if (password.length > 4) {
        // Password has to be at least 4 characters long
        this.setState({validPassword: true});
      }
    } else if (password <= 4) {
      this.setState({validPassword: false});
    }
  }

  handleCloseNotification() {
    this.setState({formValid: true});
  }

  toggleNextButtonState() {
    const {validEmail, validPassword} = this.state;
    if (validEmail && validPassword) {
      return false;
    }
    return true;
  }

  render() {
    const {formValid, loadingVisible, validEmail, validPassword} = this.state;
    const showNotification = !formValid;
    const background = formValid ? colors.green01 : colors.darkOrange;
    const notificationMarginTop = showNotification ? 10 : 0;
    return (
      <KeyboardAvoidingView
        style={[{backgroundColor: background}, styles.wrapper]}
        behavior="padding">
        <View style={styles.scrollViewWrapper}>
          <ScrollView style={styles.scrollView}>
            <Text style={styles.loginHeader}>Log In</Text>
            <InputField
              labelText="EMAIL ADDRESS"
              labelTextSize={14}
              labelColor={colors.white}
              textColor={colors.white}
              borderBottomColor={colors.white}
              inputType="email"
              customStyle={{marginBottom: 30}}
              onChangeText={this.handleEmailChange}
              showCheckMark={validEmail}
              autoFocus={true}
            />
            <InputField
              labelText="PASSWORD"
              labelTextSize={14}
              labelColor={colors.white}
              textColor={colors.white}
              borderBottomColor={colors.white}
              inputType="password"
              customStyle={{marginBottom: 30}}
              onChangeText={this.handlePasswordChange}
              showCheckMark={validPassword}
            />
          </ScrollView>
          <View style={styles.nextButton}>
            <NextArrowButton
              handleNextButton={this.handleNextButton}
              disabled={this.toggleNextButtonState()}
            />
          </View>
        </View>
        <View
          style={[
            styles.notificationWrapper,
            {marginTop: notificationMarginTop},
          ]}>
          <Notification
            showNotification={showNotification}
            handleCloseNotification={this.handleCloseNotification}
            type="Error"
            firstLine="Those credentials don't look right."
            secondLine="Please try again."
          />
        </View>
        {/* <Loader modalVisible={loadingVisible} animationType="fade" /> */}
      </KeyboardAvoidingView>
    );
  }
}

Login.propTypes = {
  login: PropTypes.func.isRequired,
  navigation: propTypesNavigation,
};

export default Login;
