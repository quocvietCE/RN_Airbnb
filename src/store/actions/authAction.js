import {AUTH_ACTION} from '../actionTypes';

const login = (email, password) => ({
  type: AUTH_ACTION.USER_LOGIN,
  payload: {
    email,
    password,
  },
});

const authAction = {
  login,
};

export default authAction;
