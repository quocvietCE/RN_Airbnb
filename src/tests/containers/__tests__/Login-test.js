import 'react-native';
import expect from 'expect';
import React from 'react';
import Login from '../../../containers/login';
import {Provider} from 'react-redux';
import renderer from 'react-test-renderer';
import Adapter from 'enzyme-adapter-react-16';
import configureStore from 'redux-mock-store';
import InputField from '../../../components/form/InputField';
import authAction from '../../../store/actions/authAction';
import Enzyme, {mount, shallow} from 'enzyme';

Enzyme.configure({adapter: new Adapter()});

const middlewares = []; // you can mock any middlewares here if necessary

const mockStore = configureStore(middlewares);

const initialState = {
  auth: {
    loggedInState: false,
  },
};

// const store = mockStore(initialState);

describe('Test Login container', () => {
  let store;
  let component;
  let wrapper;
  let wrapperShallow;
  const mockLoginfn = jest.fn();
  const defaultState = {
    formValid: true,
    validEmail: false,
    emailAddress: '',
    password: '',
    validPassword: false,
    loadingVisible: false,
    secureInput: 'text',
  };

  beforeEach(() => {
    //creates the store with any initial state or middleware needed
    store = mockStore(initialState);
    store.dispatch = jest.fn();
    component = renderer.create(
      <Provider store={store}>
        <Login />
      </Provider>,
    );
    // wrapper = mount(
    //   <Provider store={store}>
    //     <Login />
    //   </Provider>,
    // );
    wrapperShallow = shallow(<Login store={store} login={mockLoginfn} />);
  });

  afterEach(() => {
    component.unmount();
  });

  it('should render with given state from Redux store', () => {
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('should dispatch an action on button click', () => {
    renderer.act(() => {
      component.root.findByType('button').props.onClick();
    });
    renderer.act(() => {
      component.root
        .findByType('input')
        .props.onChange({target: {value: 'some other text'}});
    });
    expect(store.dispatch).toHaveBeenCalledTimes(1);
    expect(store.dispatch).toHaveBeenCalledWith(
      authAction.login('test@test.com', '123456'),
    );
  });

  it('renders the inner InputField', () => {
    expect(wrapperShallow.find(InputField).length).toEqual(2);
  });
});
