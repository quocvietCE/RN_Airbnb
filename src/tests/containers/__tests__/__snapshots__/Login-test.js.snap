// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Login container should render with given state from Redux store 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Array [
        Object {
          "backgroundColor": "#008388",
        },
        Object {
          "display": "flex",
          "flex": 1,
        },
      ],
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <View
    style={
      Object {
        "bottom": 0,
        "flex": 1,
        "left": 0,
        "marginTop": 70,
        "padding": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      }
    }
  >
    <RCTScrollView
      style={
        Object {
          "flex": 1,
          "paddingLeft": 30,
          "paddingRight": 30,
          "paddingTop": 20,
        }
      }
    >
      <View>
        <Text
          style={
            Object {
              "color": "#ffffff",
              "fontSize": 34,
              "fontWeight": "300",
              "marginBottom": 40,
            }
          }
        >
          Log In
        </Text>
        <View
          style={
            Array [
              Object {
                "marginBottom": 30,
              },
              Object {
                "display": "flex",
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#ffffff",
                  "fontSize": 14,
                  "fontWeight": "700",
                },
                Object {
                  "marginBottom": 20,
                },
              ]
            }
          >
            EMAIL ADDRESS
          </Text>
          <View
            style={
              Object {
                "bottom": 12,
                "position": "absolute",
                "right": 0,
                "transform": Array [
                  Object {
                    "scale": 0.01,
                  },
                ],
              }
            }
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#ffffff",
                    "fontSize": 20,
                  },
                  undefined,
                  Object {
                    "fontFamily": "FontAwesome",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
          </View>
          <TextInput
            allowFontScaling={true}
            autoCorrect={false}
            autoFocus={true}
            keyboardType="email-address"
            onChangeText={[Function]}
            rejectResponderTermination={true}
            secureTextEntry={false}
            selectionColor="white"
            style={
              Array [
                Object {
                  "borderBottomColor": "#ffffff",
                  "color": "#ffffff",
                },
                undefined,
                Object {
                  "borderBottomWidth": 1,
                  "paddingTop": 5,
                },
              ]
            }
            underlineColorAndroid="transparent"
          />
        </View>
        <View
          style={
            Array [
              Object {
                "marginBottom": 30,
              },
              Object {
                "display": "flex",
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#ffffff",
                  "fontSize": 14,
                  "fontWeight": "700",
                },
                Object {
                  "marginBottom": 20,
                },
              ]
            }
          >
            PASSWORD
          </Text>
          <View
            accessible={true}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
                "position": "absolute",
                "right": 0,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#ffffff",
                  "fontWeight": "700",
                }
              }
            >
              Show
            </Text>
          </View>
          <View
            style={
              Object {
                "bottom": 12,
                "position": "absolute",
                "right": 0,
                "transform": Array [
                  Object {
                    "scale": 0.01,
                  },
                ],
              }
            }
          >
            <Text
              allowFontScaling={false}
              style={
                Array [
                  Object {
                    "color": "#ffffff",
                    "fontSize": 20,
                  },
                  undefined,
                  Object {
                    "fontFamily": "FontAwesome",
                    "fontStyle": "normal",
                    "fontWeight": "normal",
                  },
                  Object {},
                ]
              }
            >
              
            </Text>
          </View>
          <TextInput
            allowFontScaling={true}
            autoCorrect={false}
            keyboardType="default"
            onChangeText={[Function]}
            rejectResponderTermination={true}
            secureTextEntry={true}
            selectionColor="white"
            style={
              Array [
                Object {
                  "borderBottomColor": "#ffffff",
                  "color": "#ffffff",
                },
                undefined,
                Object {
                  "borderBottomWidth": 1,
                  "paddingTop": 5,
                },
              ]
            }
            underlineColorAndroid="transparent"
          />
        </View>
      </View>
    </RCTScrollView>
    <View
      style={
        Object {
          "alignItems": "flex-end",
          "bottom": 20,
          "position": "absolute",
          "right": 20,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "flex-end",
            "bottom": 20,
            "paddingTop": 5,
            "right": 20,
          }
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "opacity": 0.2,
              },
              Object {
                "alignItems": "center",
                "backgroundColor": "#ffffff",
                "borderRadius": 50,
                "height": 50,
                "justifyContent": "center",
                "width": 50,
              },
            ]
          }
        >
          <Text
            allowFontScaling={false}
            style={
              Array [
                Object {
                  "color": "#008388",
                  "fontSize": 32,
                },
                Object {
                  "marginRight": -2,
                  "marginTop": -2,
                },
                Object {
                  "fontFamily": "FontAwesome",
                  "fontStyle": "normal",
                  "fontWeight": "normal",
                },
                Object {},
              ]
            }
          >
            
          </Text>
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "zIndex": 9,
        },
        Object {
          "marginTop": 0,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#ffffff",
          "flex": 1,
          "height": 80,
          "padding": 10,
          "transform": Array [
            Object {
              "translateY": -80,
            },
          ],
        }
      }
    >
      <View
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
            "marginBottom": 2,
          }
        }
      >
        <View
          style={
            Array [
              Object {
                "flex": 1,
                "flexDirection": "row",
                "fontSize": 14,
                "marginBottom": 2,
              },
            ]
          }
        >
          <Text
            style={
              Array [
                Object {
                  "color": "#d93900",
                  "fontSize": 14,
                  "marginBottom": 2,
                  "marginRight": 5,
                },
              ]
            }
          >
            Error
          </Text>
          <Text
            style={
              Object {
                "flex": 1,
                "flexDirection": "row",
                "fontSize": 14,
                "marginBottom": 2,
              }
            }
          >
            Those credentials don't look right.
             
            Please try again.
             
          </Text>
        </View>
      </View>
      <View
        accessible={true}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
            "position": "absolute",
            "right": 10,
            "top": 10,
          }
        }
      >
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "color": "#d8d8d8",
                "fontSize": 20,
              },
              undefined,
              Object {
                "fontFamily": "FontAwesome",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          
        </Text>
      </View>
    </View>
  </View>
</View>
`;
